pkg/providers/vsphere/upgrade/virtualmachine/vm_schema_upgrade.go

```diff
@@ -397,39 +406,358 @@ func reconcileSCSIController(
+		ctx, k8sClient, vm)
+	if err != nil {
+		return
+	}
+
+	for i := range vm.Spec.Volumes {
+		vol := &vm.Spec.Volumes[i]
+
+		if vol.PersistentVolumeClaim == nil {
+			continue
 		}
+
+		reconcilePVCVolumePlacement(
+			ctx,
+			vol,
+			volumeNameToDiskUUID,
+			diskUUIDToDiskInfo,
+			ctrlKeyToCtrlStatus)
 	}
 }
 
-func reconcileVirtualDisk(
+// buildDiskUUIDToDiskInfoMap builds a map of disk UUID to disk
+// device info by extracting from vSphere VM hardware.
+func buildDiskUUIDToDiskInfoMap(
+	diskDevices []*vimtypes.VirtualDisk) map[string]pkgutil.VirtualDiskInfo {
+
+	diskUUIDToDiskInfo := make(map[string]pkgutil.VirtualDiskInfo)
+	for _, disk := range diskDevices {
+		vdi := pkgutil.GetVirtualDiskInfo(disk)
+		if vdi.UUID != "" {
+			diskUUIDToDiskInfo[vdi.UUID] = vdi
+		}
+	}
+	return diskUUIDToDiskInfo
+}
+
+// buildVolumeNameToDiskUUIDMap builds a map of volume name to
+// disk UUID by querying CnsNodeVmAttachment objects. This is the
+// source of truth for PVC volume attachments.
+func buildVolumeNameToDiskUUIDMap(
 	ctx context.Context,
-	_ *vmopv1.VirtualMachine,
-	_ *vimtypes.VirtualDisk) {
+	k8sClient ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) (map[string]string, error) {
 
 	logger := pkglog.FromContextOrDefault(ctx)
-	logger.V(4).Info("Reconciling schema upgrade for VM disk")
+	volumeNameToDiskUUID := make(map[string]string)
+	attachments, err := pkgutil.GetAttachmentsForVM(ctx, k8sClient, vm)
+	if err != nil {
+		logger.V(4).Info(
+			"Failed to get CnsNodeVmAttachments", "error", err)
+		return nil, err
+	}
+
+	for _, attachment := range attachments {
+		if attachment.Status.Attached {
+			diskUUID := attachment.Status.AttachmentMetadata[cnsv1alpha1.AttributeFirstClassDiskUUID]
+			if diskUUID != "" {
+				volumeNameToDiskUUID[attachment.Spec.VolumeName] =
+					diskUUID
+			}
+		}
+	}
+	return volumeNameToDiskUUID, nil
+}
+
+// reconcilePVCVolumePlacement reconciles the placement info for
+// a single PVC volume.
+func reconcilePVCVolumePlacement(
+	ctx context.Context,
+	vol *vmopv1.VirtualMachineVolume,
+	volumeNameToDiskUUID map[string]string,
+	diskUUIDToDiskInfo map[string]pkgutil.VirtualDiskInfo,
+	ctrlKeyToCtrlStatus map[int32]vmopv1.VirtualControllerStatus) {
+
+	logger := pkglog.FromContextOrDefault(ctx)
+	pvc := vol.PersistentVolumeClaim
+
+	diskUUID, ok := volumeNameToDiskUUID[pvc.ClaimName]
+	if !ok {
+		logger.V(4).Info(
+			"Skipping PVC volume without disk UUID",
+			"volume", vol.Name)
+		return
+	}
+
+	info, ok := diskUUIDToDiskInfo[diskUUID]
+	if !ok {
+		logger.V(4).Info(
+			"Disk UUID is not found in VM hardware",
+			"volume", vol.Name,
+			"diskUUID", diskUUID)
+		return
+	}
 
+	ctrl, ok := ctrlKeyToCtrlStatus[info.ControllerKey]
+	if !ok {
+		logger.V(4).Info(
+			"Controller not found in status for PVC volume",
+			"volume", vol.Name,
+			"controllerKey", info.ControllerKey,
+			"diskUUID", diskUUID)
+		return
+	}
+
+	if !needsPlacementBackfill(pvc) {
```

Comment:
Is it clearer to do this first thing in this function?

Reply:
yeah, moved. just realized that after a self-review too lol